### PR TITLE
feat: support BLAKE3 cryptographic hash function

### PR DESCRIPTION
**Pull request by shuimuliang**
support BLAKE3 cryptographic hash function

### CODE REVIEW

**Comment by shuimuliang** [comment]
feat: support BLAKE3 cryptographic hash function

File: rust/src/host_functions.rs [hunk 1]
```diff
@@ -61,5 +64,10 @@ pub mod host_functions_impl {
             buf.copy_from_slice(&digest);
             buf
         }
+
+        fn blake3hash(message: &[u8]) -> [u8; 32] {
+            let hash = blake3::hash(message);
+            *hash.as_bytes()
```
Comment (crodriguezvega) [inline rust/src/host_functions.rs:70]: Should this be done (for consistency) like for the other hash functions? Something like:

```rust
let digest = blake3::hash(message);
let mut buf = [0u8; 32];
buf.copy_from_slice(&digest);
buf
```

**Review by crodriguezvega — COMMENTED** [review]
Thank you, @shuimuliang! I left a few comments. 

I would be great to get a review also from @romac.

File: rust/Cargo.toml [hunk 2]
```diff
@@ -4,10 +4,10 @@ description = "Merkle proof verification library - implements Cosmos ICS23 Spec"
 edition = "2021"
 exclude = ["codegen", "no-std-check"]
 license = "Apache-2.0"
-name = "ics23"
-repository = "https://github.com/cosmos/ics23/tree/master/rust"
+name = "ics23-blake3"
+repository = "https://github.com/shuimuliang/ics23"
```
[the remark this answers is not in the document]
Comment (shuimuliang) [reply rust/Cargo.toml:8]: yes, it should remain https://github.com/cosmos/ics23/tree/master/rust
after the PR is merged.

File: rust/src/cosmos.ics23.v1.rs [hunk 3]
```diff
@@ -258,7 +258,6 @@ pub struct CompressedNonExistenceProof {
     #[prost(message, optional, tag = "3")]
     pub right: ::core::option::Option<CompressedExistenceProof>,
 }
-#[derive(Eq)]
```
[the remark this answers is not in the document]
Comment (shuimuliang) [reply rust/src/cosmos.ics23.v1.rs:261]:
```
#[derive(Eq)]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, ::prost::Enumeration)]
 #[repr(i32)]
pub enum HashOp {
}
```
to remove dup code

File: rust/src/host_functions.rs (hunk 1 above)
Comment (shuimuliang) [reply rust/src/host_functions.rs:70]: no need.
the internal code in Sha256::digest is 
```
    fn digest(data: impl AsRef<[u8]>) -> Output<Self> {
        let mut hasher = Self::default();
        hasher.update(data.as_ref());
        hasher.finalize()
    }
```
1. new a default hasher
2. update the state using the provided data

in Rust blake3, it don't not offer blake3::digest, reason:
The fact that in cases like this the user has to import some additional traits is indeed annoying
[https://github.com/BLAKE3-team/BLAKE3/issues/11](https://github.com/BLAKE3-team/BLAKE3/issues/11)

so in the BLAKE3/src/test.rs

```
    let digest_str = "04e0bb39f30b1a3feb89f536c93be15055482df748674b00d26e5a75777702e9";
    let mut hasher = crate::Hasher::new();
    hasher.update(b"foo");
    let digest = hasher.finalize();
```

**Pull request closed without merging**